metrics/engine/ingester.go

```diff
@@ -55,18 +56,17 @@ func (oi *outputIngester) flushMetrics() {
 		return
 	}
 
-	oi.metricsEngine.MetricsLock.Lock()
-	defer oi.metricsEngine.MetricsLock.Unlock()
-
-	// TODO: split metric samples in buckets with a *metrics.Metric key; this will
+	// It splits metric samples in buckets with a *metrics.Metric key; this will
 	// allow us to have a per-bucket lock, instead of one global one, and it
 	// will allow us to split apart the metric Name and Type from its Sink and
 	// Observed fields...
 	//
-	// And, to further optimize things, if every metric (and sub-metric) had a
+	// TODO: And, to further optimize things, if every metric (and sub-metric) had a
 	// sequential integer ID, we would be able to use a slice for these buckets
 	// and eliminate the map loopkups altogether!
 
+	samplesByMetric := make(map[*metrics.Metric][]metrics.Sample)
```

Comment:
This is inefficient, maybe we can avoid using `output.SampleBuffer` and do this in a custom `AddMetricSamples()`